src/ngraph/ngraph_nnvm_ops.cc

```diff
@@ -94,22 +99,30 @@ void compute_backward(const mxnet::OpContext &ctx, std::shared_ptr<Graph> graph,
   // generate valid data in fprop cache
   if (graph->enable_fprop_cache && !graph->forward_train_computed) {
     // forward inputs
-    std::vector<mxnet::NDArray> fwd_inputs(inputs.begin() + graph->num_outputs,
+    std::vector<mxnet::NDArray> fwd_inputs(inputs.begin() + graph->num_outputs_,
                                            inputs.end());
     auto placeholders = make_ngraph_placeholders(fwd_inputs, backend, true);
     // forward outputs
-    auto shape = TShape_to_NShape(graph->nodes_.back()->shape_);
-    const auto &element_type = getType(graph->nodes_.back()->dtype_);
-    auto output_tv = backend->make_primary_tensor_view(element_type, shape);
-    TensorViewVector results{output_tv};
+    TensorViewVector results;
+    for (size_t i = 0; i < graph->num_outputs_; ++i) {
+      auto shape = TShape_to_NShape(graph->outputs_[i]->shape_);
+      const auto &element_type = getType(graph->outputs_[i]->dtype_);
+      auto output_tv = backend->make_primary_tensor_view(element_type, shape);
+      results.push_back(output_tv);
+    }
     append_cached_to_forward(&results, graph, mode);
     // call forward
     graph->ngraph_forward[mode]->call(results, placeholders);
   }
 
   // backward op
+  std::vector<mxnet::NDArray> adjoints;
+  for (size_t i = 0; i < graph->num_outputs_; ++i) {
+    adjoints.push_back(inputs[i]);
+  }
```

Comment:
My comment above regarding how to concatenate vectors may apply here as well.

Reply:
Inputs will generally be longer than adjoints in a non-trivial way.

Reply:
Cool.  IMO it would be nice to have a brief comment mentioning that.  Or you could say the same thing more concisely with:
``` c++
CHECK(inputs.size() >= graph->num_outputs_);
```

Up to you, of course.

Reply:
I think the for loop is easier to understand in this case.